core/blockchain.go

```diff
@@ -250,20 +254,51 @@ func (bc *BlockChain) VerifyNonce(accountIndex int64, nonce int64) error {
 			return err
 		}
 		if nonce != expectNonce {
-			return errors.New("invalid Nonce")
+			return errors.New("invalid nonce")
 		}
 	} else {
 		pendingNonce, err := bc.Statedb.GetPendingNonce(accountIndex)
 		if err != nil {
 			return err
 		}
 		if pendingNonce != nonce {
-			return errors.New("invalid Nonce")
+			return errors.New("invalid nonce")
 		}
 	}
 	return nil
 }
 
+func (bc *BlockChain) VerifyGas(accountIndex, gasAccountIndex, gasFeeAssetId int64, gasFeeAssetAmount *big.Int) error {
+	gasAccountConfig, err := bc.ChainDB.SysConfigModel.GetSysConfigByName(types2.GasAccountIndex)
+	if err != nil {
+		logx.Errorf("cannot find config for: %s", types2.GasAccountIndex)
+		return errors.New("internal error")
+	}
+	index, err := strconv.ParseInt(gasAccountConfig.Value, 10, 64)
+	if err != nil {
+		logx.Errorf("invalid account index: %s", gasAccountConfig.Value)
+		return errors.New("internal error")
+	}
+	if gasAccountIndex != index {
+		return errors.New("invalid gas fee account")
+	}
+
+	gasAssetConfig, err := bc.ChainDB.L2AssetInfoModel.GetAssetById(gasFeeAssetId)
+	if err != nil {
+		logx.Errorf("cannot find gas asset: %s", err.Error())
+		return errors.New("invalid gas fee asset")
+	}
+	if gasAssetConfig.IsGasAsset != asset.IsGasAsset {
+		return errors.New("invalid gas fee asset")
+	}
+
+	accountGasAsset := bc.Statedb.AccountMap[accountIndex].AssetInfo[gasFeeAssetId]
```

Comment:
The gas amount will be compared in each tx executor's verifyInput function, I think this is redundant here. We can just verify the gasAccount and gasFeeAsset here because the balance may be mixed by gas and other assets, only compare gas fee asset amount isn't enough.

Reply:
For BSC, it will sub the balance of the account for gas fee, here the logic about balance and fee is deeply 
coupled, very easy to have bugs.

Reply:
removed gas amount check